sapicore/pipeline/simulation.py

```diff
@@ -0,0 +1,209 @@
+from torch import Tensor
+
+from alive_progress import alive_bar
+
+from sapicore.pipeline import Pipeline
+from sapicore.model import Model
+from sapicore.engine.network import Network
+
+from sapicore.utils.seed import fix_random_seed
+from sapicore.utils.signals import extend_input_current
+from sapicore.utils.tensorboard import TensorboardWriter, HDFData
+
+from sapicore.utils.constants import DT, TIME_FORMAT, SEED
+from sapicore.utils.io import ensure_dir, save_yaml, log_settings
+
+from sapicore.tests import ROOT
+
+__all__ = ("Simulator",)
+
+
+class Simulator(Pipeline):
+    """A prototypical simulation pipeline, bringing together model specification, disk I/O, and visualization.
+    Initialize a network and configure the simulation by providing a dictionary or a path to a YAML.
+
+    Parameters
+    ----------
+    configuration: dict or str
+        Configuration dictionary or path to a configuration YAML. If the path provided is relative, the simulation
+        pipeline will assume that the root is `sapicore/tests` to support the execution of YAML-configured tests.
+
+    """
+
+    def __init__(self, configuration: dict | str, **kwargs):
+        # loads the configuration if a path is provided.
+        super().__init__(configuration=configuration, **kwargs)
+
+        if self.configuration and self.configuration.get("device"):
+            # set hardware device from configuration.
+            self.device = self.configuration.get("device")
+
+        else:
+            # set default hardware device for this pipeline.
+            self.device = "cpu" if not torch.cuda.is_available() else "cuda:0"
+
+        self.initialize(**kwargs)
+
+    def initialize(self, seed: int = None):
+        """Initializes the simulation pipeline. Validates project root path, sets random number generation seed,
+        and applies log message configuration.
+
+        Parameters
+        ----------
+        seed: int, optional
+            Random seed to be used throughout the simulation. Applied to all relevant libraries.
+            If not provided, looks up "seed" in the configuration dictionary.
+            If not found, the project default :attr:`utils.constants.SEED` is used to ensure reproducibility.
+
+        Warning
+        -------
+        Deterministic behavior is vital when conducting reproducible experiments. This is one of many safeguards
+        included in Sapicore. Users who wish to modify this default behavior will have to consciously override
+        :meth:`~Simulator.initialize`.
+
+        """
+        # account for the relative path case, e.g. if invoked by pytest.
+        project_root = self.configuration.get("root")
+        if not project_root:
+            raise KeyError("Configuration missing project 'root' key.")
+
+        # account for the relative path case, e.g. if invoked by pytest.
+        if not os.path.isabs(project_root):
+            self.configuration["root"] = os.path.join(ROOT, project_root)
```

Comment:
It's been a headache because of the competing constraints: (1) tests use YAMLs, so I had to support relative paths (can't force using abs paths and still provide YAMLs that work out of the box) (2) pytest acts up if you don't fix the path (3) simulation.py must interact with the test directory if used in a test, but can otherwise accept any YAML containing a (project) `root` key and run with it.

>  bad sign of eventual circular dependencies and/or of isolation issues

To me, this suggests that simulation.py should be banished to the test suite. It has been extremely useful during development/testing, but now that we have multiple examples of how to write an end-to-end pipeline (basic_training and sapinet-neurips), we should probably break it down and send the pieces where they belong (e.g., current design to data synthesis, tensorboard to plotting, etc.).

Reply:
Agreeing with this although I would do that in one of the next PR.

Reply:
Ended up fixing the path issues and expanding simulation.py (now generic.py and `GenericSimulator` to convey that it's just one point in a vast `Pipeline` space). It now supports passing `Data` or a tensor. If no data is provided, it will pass zeros (possibly + a constant `current`) for a configured duration `steps`, as usual. The temporary `artifical_current` function was scrapped.